xla/service/cpu/onednn_ops_rewriter.cc

```diff
@@ -143,183 +139,288 @@ std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
+                            reduce->dimensions().size() == 1 &&
+                            reduce->dimensions()[0] == reduce->shape().rank());
+                  }))
+          .WithOperand(1, m::Op(&broadcast1).WithOpcode(HloOpcode::kBroadcast));
+
+  if (!Match(div1, div_red_subgraph)) return false;
+
+  if (broadcast1 != broadcast0 || reduce_in0 != mul_in0 || mul_in0 != prod_s) {
+    return false;
+  }
+
+  *is_producer_bf16orfp16 =
+      (prod_s->shape().element_type() == PrimitiveType::F16) ||
+      (prod_s->shape().element_type() == PrimitiveType::BF16);
+  if (instr->user_count() == 1 &&
+      instr->users().at(0)->opcode() == HloOpcode::kConvert) {
+    *convert_instr = instr->users().at(0);
+    *is_bf16orfp16_convert =
+        ((*convert_instr)->shape().element_type() == PrimitiveType::F16 ||
+         (*convert_instr)->shape().element_type() == PrimitiveType::BF16);
+  }
+
+  *src = prod_s;
+  *scale = scale_gamma;
+  *bias = shift;
+  // get epsilon
+  *eps = static_cast<float>(epsilon->literal().GetAsDouble({}).value());
+
+  return true;
+}
+
 }  // namespace
 
 class OneDnnOpsRewriterVisitor : public DfsHloRewriteVisitor {
  public:
   Status HandleAdd(HloInstruction* instr) override {
-    HloInstruction *slicemu1, *slicemu2;
-    HloInstruction *slicesource1, *slicesource2;
-    HloInstruction *musquare1, *musquare2;
-    HloInstruction *prod_c, *prod_l, *prod_s, *prod_r;
-    HloInstruction *slicevar, *hinge;
-
-    bool scaleFound = false;
-    bool shiftFound = false;
-
-    auto spine = m::Add().WithBinaryOperandsAnyOrder(
-        m::Broadcast(),
-        m::Multiply()
-            .WithBinaryOperandsAnyOrder(
-                m::Op(&hinge).WithOneUser(),
-                m::Subtract(
-                    m::Op(&prod_s),
-                    m::Broadcast(
-                        m::Reshape(
-                            m::Broadcast(
-                                m::Reshape(
-                                    m::Op(&slicemu1)
-                                        .WithOpcode(HloOpcode::kSlice)
-                                        .WithOperand(
-                                            0, m::Op(&slicesource1)
-                                                   .WithOpcode(
-                                                       HloOpcode::kDivide)))
-                                    .WithOneUser())
-                                .WithOneUser())
-                            .WithOneUser())
-                        .WithOneUser())
-                    .WithOneUser())
-            .WithOneUser());
-
-    if (!Match(instr, spine)) {
-      return OkStatus();
+    HloInstruction *src, *scale, *bias;
+    float eps;
+    bool found_ln = false;
+    bool is_bf16orfp16_convert = false;
+    bool is_producer_bf16orfp16 = false;
+    HloInstruction* convert_instr;
+
+    if (MatchTFKerasLayerNorm(instr, &src, &scale, &bias, &eps)) {
+      found_ln = true;
     }
```

Comment:
Nit: This can be simplified to
```suggestion
    bool found_ln = MatchTFKerasLayerNorm(instr, &src, &scale, &bias, &eps));
```

Reply:
Done